plugins/file-open-save-new/core/src/test/java/org/pentaho/di/plugins/fileopensave/providers/vfs/VFSFileProviderTest.java

```diff
@@ -0,0 +1,215 @@
+
+    assertTrue( testInstance.isConnectionRoot( createTestInstance( "pvfs://someConnection123" ) ) );
+
+    assertTrue( testInstance.isConnectionRoot( createTestInstance( "pvfs://123someConnection123" ) ) );
+
+    assertTrue( testInstance.isConnectionRoot( createTestInstance( "pvfs://someConnection/" ) ) );
+
+    assertTrue( testInstance.isConnectionRoot( createTestInstance( "pvfs://some_Connection/" ) ) );
+
+    assertTrue( testInstance.isConnectionRoot( createTestInstance( "pvfs://some-Connection/" ) ) );
+
+    assertTrue( testInstance.isConnectionRoot( createTestInstance( "pvfs://someConnection123/" ) ) );
+
+    assertTrue( testInstance.isConnectionRoot( createTestInstance( "pvfs://123someConnection123/" ) ) );
+
+    // TEST now we have past the root domain
+    assertFalse( testInstance.isConnectionRoot( createTestInstance( "pvfs://someConnection/someFolderA" ) ) );
+
+    assertFalse( testInstance.isConnectionRoot(
+        createTestInstance( "pvfs://someConnection/someFolderA/directory2" ) ) );
+
+    assertFalse( testInstance.isConnectionRoot(
+        createTestInstance( "pvfs://someConnection/someFolderA/directory2/randomFileC.txt" ) ) );
+  }
+
+  public void testHasBuckets() {
+
+    VFSFileProvider testInstance = createTestInstance();
+
+    // TEST non VFSConnectionDetails
+    ConnectionDetails mockConnectionDetails  = mock( ConnectionDetails.class );
+    assertFalse( testInstance.hasBuckets( mockConnectionDetails ) );
+
+    // TEST  does not have buckets
+    VFSConnectionDetails mock_NoBuckets_VFSConnectionDetails  = mock( VFSConnectionDetails.class );
+    when( mock_NoBuckets_VFSConnectionDetails.hasBuckets() ).thenReturn( false );
+    assertFalse( testInstance.hasBuckets( mock_NoBuckets_VFSConnectionDetails ) );
+
+    // TEST has buckets
+    VFSConnectionDetails mock_Buckets_VFSConnectionDetails  = mock( VFSConnectionDetails.class );
+    when( mock_Buckets_VFSConnectionDetails.hasBuckets() ).thenReturn( true );
+    assertTrue( testInstance.hasBuckets( mock_Buckets_VFSConnectionDetails ) );
+  }
+
+  @Test
+  public void testIsSupported() throws Exception {
+
+    VFSFileProvider testInstance = createTestInstance();
+
+    assertFalse( testInstance.isSupported( null ) );
+
+    assertFalse( testInstance.isSupported( "" ) );
+
+    assertFalse( testInstance.isSupported( "    " ) );
+
+    assertFalse( testInstance.isSupported( "someGarbage" ) );
+
+    assertFalse( testInstance.isSupported( "/someUser/someUnixFile" ) );
+
+    assertFalse( testInstance.isSupported( "T:\\Users\\RandomSUser\\Documents\\someWindowsFile" ) );
+
+    assertFalse( testInstance.isSupported( "//home/randomUser/randomFile.rpt" ) );
+
+    assertTrue( testInstance.isSupported( "pvfs://someConenction/someFilePath" ) );
+  }
+
+  @Test
+  public void testGetConnectionName() throws Exception {
+
+    VFSFileProvider testInstance = createTestInstance();
```

Comment:
Should include tests with connection names with special characters.

Reply:
You can create a VFS connection with special characters such as "###" why, Idk. But I should be in sync with the other parts of the UI. Let me see if I can find the code that puts any restrictions on the connection name, otherwise, I'll just allow anything on my part and consider it the duty of other parts of the code to manage the allowed characters.

Reply:
Addressed see commit https://github.com/pentaho/pentaho-kettle/commit/3e8bcc9f08d5634a303f5bdcbaf73f80d1811e46